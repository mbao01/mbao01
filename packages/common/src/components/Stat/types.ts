import React from "react";
import { type VariantProps } from "../../libs";
import {
  getStatActionsClasses,
  getStatClasses,
  getStatDescriptionClasses,
  getStatFigureClasses,
  getStatsClasses,
  getStatTitleClasses,
  getStatValueClasses,
} from "./constants";

export type StatsProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getStatsClasses>;

export type StatProps = React.HTMLAttributes<HTMLDivElement> & VariantProps<typeof getStatClasses>;

export type StatActionsProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getStatActionsClasses>;

export type StatDescriptionProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getStatDescriptionClasses>;

export type StatFigureProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getStatFigureClasses>;

export type StatTitleProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getStatTitleClasses>;

export type StatValueProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getStatValueClasses>;
