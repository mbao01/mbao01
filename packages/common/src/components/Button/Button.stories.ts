import type { Meta, StoryObj } from '@storybook/react';

import { Button } from './Button';

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: 'centered',
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    variant: 'default',
    label: 'Button',
  },
};

export const PrimaryButton: Story = {
  args: {
    variant: 'primary',
    label: 'Button',
  },
};

export const WideButton: Story = {
  args: {
    wide: true,
    label: 'Wide Button',
  },
};

export const TinyButton: Story = {
  args: {
    size: 'xs',
    label: 'Tiny',
  },
};

export const OutlineButton: Story = {
  args: {
    outline: true,
    label: 'Click me!',
  },
};

export const DisabledButton: Story = {
  args: {
    disabled: true,
    label: 'Click me?',
  },
};

export const LoadingButton: Story = {
  args: {
    loading: true,
    label: 'Loading',
  },
};

export const BlockButton: Story = {
  args: {
    disabled: true,
    className: 'btn-block',
    label: 'Block Button',
  },
};
