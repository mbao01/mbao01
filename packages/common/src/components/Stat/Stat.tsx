import type {
  StatActionsProps,
  StatDescriptionProps,
  StatFigureProps,
  StatProps,
  StatsProps,
  StatTitleProps,
  StatValueProps,
} from "./types";
import { cn } from "../../utilities";
import {
  getStatActionsClasses,
  getStatClasses,
  getStatDescriptionClasses,
  getStatFigureClasses,
  getStatsClasses,
  getStatTitleClasses,
  getStatValueClasses,
} from "./constants";

const Stats = ({ direction, className, ...props }: StatsProps) => (
  <div className={cn(getStatsClasses({ direction }), className)} {...props} />
);

const Stat = ({ className, position, ...props }: StatProps) => (
  <div className={cn(getStatClasses({ position }), className)} {...props} />
);

const StatTitle = ({ className, ...props }: StatTitleProps) => (
  <div className={cn(getStatTitleClasses(), className)} {...props} />
);

const StatValue = ({ className, ...props }: StatValueProps) => (
  <div className={cn(getStatValueClasses(), className)} {...props} />
);

const StatDescription = ({ className, ...props }: StatDescriptionProps) => (
  <div className={cn(getStatDescriptionClasses(), className)} {...props} />
);

const StatFigure = ({ className, ...props }: StatFigureProps) => (
  <div className={cn(getStatFigureClasses(), className)} {...props} />
);

const StatActions = ({ className, ...props }: StatActionsProps) => (
  <div className={cn(getStatActionsClasses(), className)} {...props} />
);

Stat.Actions = StatActions;
Stat.Description = StatDescription;
Stat.Figure = StatFigure;
Stat.Title = StatTitle;
Stat.Value = StatValue;

export { Stats, Stat };
