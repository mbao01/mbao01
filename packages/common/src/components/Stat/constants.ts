import { cva } from "../../libs";

export const getStatsClasses = cva("stats", {
  variants: {
    direction: {
      horizontal: "stats-horizontal",
      vertical: "stats-vertical",
    },
  },
  defaultVariants: {
    direction: "horizontal",
  },
});

export const getStatClasses = cva("stat", {
  variants: {
    position: {
      center: "place-items-center",
      start: "place-items-start",
      end: "place-items-end",
    },
  },
});

export const getStatTitleClasses = cva("stat-title");

export const getStatValueClasses = cva("stat-value");

export const getStatDescriptionClasses = cva("stat-desc");

export const getStatFigureClasses = cva("stat-figure");

export const getStatActionsClasses = cva("stat-actions");
