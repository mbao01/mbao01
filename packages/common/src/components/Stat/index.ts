export { Stat, Stats } from "./Stat";
