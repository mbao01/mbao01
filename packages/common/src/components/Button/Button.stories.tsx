import type { Meta, StoryObj } from "@storybook/react";
import { Button } from "./Button";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Button,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    children: "Button",
  },
};

export const PrimaryButton: Story = {
  args: {
    variant: "primary",
    children: "Button",
  },
};

export const WideButton: Story = {
  args: {
    wide: true,
    children: "Wide Button",
  },
};

export const TinyButton: Story = {
  args: {
    size: "xs",
    children: "Tiny",
  },
};

export const OutlineButton: Story = {
  args: {
    outline: true,
    children: "Click me!",
  },
};

export const DisabledButton: Story = {
  args: {
    disabled: true,
    children: "Click me?",
  },
};

export const LoadingButton: Story = {
  args: {
    isLoading: true,
    children: "Loading",
  },
};

export const BlockButton: Story = {
  args: {
    disabled: true,
    className: "btn-block",
    children: "Block Button",
  },
};

export const SlotButton: Story = {
  args: {
    asChild: true,
    className: "btn-block",
    children: (
      <a href="#" rel="noopener">
        Link
      </a>
    ),
  },
};
