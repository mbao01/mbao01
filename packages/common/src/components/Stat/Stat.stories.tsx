import type { Meta, StoryObj } from "@storybook/react-vite";
import { Button } from "../Button";
import { Stat, Stats } from "./Stat";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Stats,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Stats>;

export default meta;
type Story = StoryObj<typeof meta>;

const StatContent = (
  <>
    <Stat.Title>Nike Shoe!</Stat.Title>
    <Stat.Value>$89,410</Stat.Value>
    <Stat.Description>↗︎ 400 (22%)</Stat.Description>
    <Stat.Actions className="gap-2">
      <Button size="xs" variant="success">
        Withdrawal
      </Button>{" "}
      <Button size="xs" name="hello" variant="secondary">
        Buy Now
      </Button>
    </Stat.Actions>
  </>
);

const StatExample = <Stat>{StatContent}</Stat>;

const CenteredStatExample = <Stat position="center">{StatContent}</Stat>;

const EndStatExample = <Stat position="end">{StatContent}</Stat>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    children: StatExample,
  },
};

export const MultipleStats: Story = {
  args: {
    children: (
      <>
        {StatExample}
        {CenteredStatExample}
        {EndStatExample}
      </>
    ),
  },
};

export const VerticalStats: Story = {
  args: {
    direction: "vertical",
    children: (
      <>
        {StatExample}
        {StatExample}
        {StatExample}
      </>
    ),
  },
};

export const BorderedStats: Story = {
  args: {
    className: "bg-base-100 border border-base-300",
    children: (
      <>
        {StatExample}
        {StatExample}
        {StatExample}
      </>
    ),
  },
  render: (args) => {
    return <Stats {...args} />;
  },
};
