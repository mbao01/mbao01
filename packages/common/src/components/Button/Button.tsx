import React from "react";
import { Slot, Slottable } from "@radix-ui/react-slot";
import { cn } from "../../utilities";
import { Loading } from "../Loading";
import { getButtonClasses } from "./constants";
import { type ButtonProps } from "./types";

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      asChild,
      className,
      outline,
      children,
      isLoading,
      variant,
      size,
      wide,
      rounded,
      disabled,
      ...props
    },
    ref
  ) => {
    const Comp = asChild ? Slot : "button";

    return (
      <Comp
        ref={ref}
        disabled={disabled}
        className={cn(
          getButtonClasses({ size, wide, outline, variant, rounded, disabled, isLoading }),
          className
        )}
        {...props}
      >
        <Slottable>{children}</Slottable>
        {isLoading ? (
          <span
            className="absolute left-0 top-0 flex h-full w-full items-center justify-center bg-[inherit]"
            data-testid="loading"
          >
            <Loading
              intent={!variant || variant === "link" || variant === "ghost" ? "default" : variant}
            />
          </span>
        ) : null}
      </Comp>
    );
  }
);

Button.displayName = "Button";

export { Button };
