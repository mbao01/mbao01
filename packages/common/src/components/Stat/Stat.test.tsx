import { render, screen } from "@testing-library/react";
import { Button } from "../Button";
import { Stat, Stats } from "./Stat";

const StatExample = (
  <Stat>
    <Stat.Title>Nike Shoe!</Stat.Title>
    <Stat.Value>$89,410</Stat.Value>
    <Stat.Description>↗︎ 400 (22%)</Stat.Description>
    <Stat.Actions className="gap-2">
      <Button size="xs" variant="success">
        Withdrawal
      </Button>{" "}
      <Button size="xs" name="hello" variant="secondary">
        Buy Now
      </Button>
    </Stat.Actions>
  </Stat>
);

describe("Stats", () => {
  it("renders a basic stats", () => {
    const { asFragment } = render(
      <Stats>
        {StatExample}
        {StatExample}
        {StatExample}
      </Stats>
    );

    const titles = screen.getAllByText("Nike Shoe!");
    titles.forEach((title) => {
      expect(title).toBeVisible();
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
