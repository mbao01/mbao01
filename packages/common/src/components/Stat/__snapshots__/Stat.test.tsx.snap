// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Stats > renders a basic stats 1`] = `
<DocumentFragment>
  <div
    class="stats stats-horizontal"
  >
    <div
      class="stat"
    >
      <div
        class="stat-title"
      >
        Nike Shoe!
      </div>
      <div
        class="stat-value"
      >
        $89,410
      </div>
      <div
        class="stat-desc"
      >
        ↗︎ 400 (22%)
      </div>
      <div
        class="stat-actions gap-2"
      >
        <button
          class="btn btn-success btn-xs rounded-md"
        >
          Withdrawal
        </button>
         
        <button
          class="btn btn-secondary btn-xs rounded-md"
          name="hello"
        >
          Buy Now
        </button>
      </div>
    </div>
    <div
      class="stat"
    >
      <div
        class="stat-title"
      >
        Nike Shoe!
      </div>
      <div
        class="stat-value"
      >
        $89,410
      </div>
      <div
        class="stat-desc"
      >
        ↗︎ 400 (22%)
      </div>
      <div
        class="stat-actions gap-2"
      >
        <button
          class="btn btn-success btn-xs rounded-md"
        >
          Withdrawal
        </button>
         
        <button
          class="btn btn-secondary btn-xs rounded-md"
          name="hello"
        >
          Buy Now
        </button>
      </div>
    </div>
    <div
      class="stat"
    >
      <div
        class="stat-title"
      >
        Nike Shoe!
      </div>
      <div
        class="stat-value"
      >
        $89,410
      </div>
      <div
        class="stat-desc"
      >
        ↗︎ 400 (22%)
      </div>
      <div
        class="stat-actions gap-2"
      >
        <button
          class="btn btn-success btn-xs rounded-md"
        >
          Withdrawal
        </button>
         
        <button
          class="btn btn-secondary btn-xs rounded-md"
          name="hello"
        >
          Buy Now
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
